Database normalization

usask-rc opened the pull request:
This will get rid of thousands of records that contain the same info, instead creating those many-to-one-to-many cross tables.

- SQL for both database platforms tested
- code tested on both my machine and frdr-dev5

Commit:
Renormallize database

- added scripts for change of schema without data loss
- changed python code to match

Commit:
Error fixes after testing new code

Commit:
One more try.

axfelix commented on harvester/DBInterface.py:
```diff
@@ -25,11 +24,11 @@ def __init__(self, params):
 			raise ValueError('Database type must be sqlite or postgres in config file')
 
+		con = self.getConnection()
 		with con:
```
I'm assuming we can do this now because the imports are handled elsewhere and this code is no longer DB interface-specific? It seems like it's partially related to your other cleanups moving out some of the DB logic, correct me if I'm wrong...

usask-rc replied:
It was just redundant to have this in two places above

axfelix commented on harvester/DBInterface.py:
```diff
@@ -38,22 +37,22 @@ def __init__(self, params):
 			# Determine if the database schema needs to be updated
 			dbversion = self.get_db_version()
-			files = os.listdir("sql/" + self.dbtype + "/")
+			files = os.listdir("sql/" + str(self.dbtype) + "/")
```
Similarly to the above -- this was working previously as far as I know, so this string recasting is just to be cleaner?

usask-rc replied:
I was accounting for the possibility that someone put a non-string value into what should be a string in the config file, which would break the code

axfelix commented on harvester/DBInterface.py:
```diff
@@ -267,41 +225,171 @@ def delete_record(self, record):
+		elif tablename == "records":
+			return "local_identifier"
+		else:
+			return tablename[:-1]
```
This is a little messy but I get what it's doing, might be better to have k:v pairs defined somewhere else if we have to differentiate "the pkey" and "the value" from the various tables more often. The `tablename[:-1]` is elegant but it seems somewhat incidental to how we've done things so far so I'm afraid it might be brittle when adding new tables if it's not more obvious elsewhere in the code...

axfelix replied:
Particularly since the `insert_related_record` function now accepts kwargs (which is also a good change)

axfelix commented on harvester/DBInterface.py:
```diff
@@ -267,41 +225,171 @@ def delete_record(self, record):
+			paramlist.append(value)
+		sqlstring1 += ")"
+		sqlstring2 += ")"
+		con = self.getConnection()
```
This is some really potentially yucky string concatentation ... I might feel better if they weren't just called `sqlstring1` and `sqlstring2` (especially since as far as I can tell they're always just whitespace separated when they're used below?)

usask-rc replied:
I agree, but the DB drivers do not allow for "?" placeholders in the table names or table column names.  Took me an hour to figure that problem out.  My previous solution was elegant but wrong.

axfelix commented on harvester/DBInterface.py:
```diff
@@ -343,198 +413,150 @@ def write_record(self, record, repo_id, metadata_prefix, domain_metadata):
+						if creator_id not in existing_creator_ids:
+							extras = {"is_contributor": 0}
+							self.insert_cross_record("records_x_creators", "creators", creator_id, record["record_id"], **extras)
 
```
Always happy to see try/except replaced with is not None tests, thanks for this

axfelix commented on harvester/DBInterface.py:
```diff
@@ -343,198 +413,150 @@ def write_record(self, record, repo_id, metadata_prefix, domain_metadata):
-									(record["record_id"], rights))
-					except self.dblayer.IntegrityError:
-						pass
+					sha1 = hashlib.sha1()
```
Wouldn't mind seeing this commented inline as to why it needs a hash for uniqueness unlike the others -- I get it but it still stands out as different

usask-rc replied:
That actually prompted me to do the same for descriptions, thanks.

axfelix commented:
OK, looks great! Made a few small comments, going to test, my only other thought is that having to maintain two separate schema files with minimal differences for SQLite and Postgres kind of sucks. Theoretically we could have one master file here and two much smaller DB-specific files that contain the` SET DEFAULT NEXTVAL` lines?

axfelix commented:
Also, let me know if you don't have time to respond to the comments today and I can just merge it after testing assuming I can't find any actual bugs, since I'll be gone for close to a week and I don't want to be a blocker, I just don't want to lose the feedback.

axfelix commented:
OK, just tested, and I think this breaks geospatial data for FGDC (tested with Polar Data Network). Otherwise looks great though! But will probably hold off on merging for that.

Commit:
Changes from feedback

- moved table columns to separate JSON file
- updated column retrieval methods
- changed descriptions table to also use a hash for lookups
- fixed bug on checking for empty dicts in geolocation and domain_metadata

usask-rc commented:
Thanks for catching the geospatial errors, that was a big miss
